Key2ds

tertsdiepraam opened the pull request.

mozzieongit commented on src/commands/key2ds.rs:
```diff
@@ -0,0 +1,166 @@
+    #[arg(short = 'n')]
+    write_to_stdout: bool,
+
+    /// use SHA1 for the DS hash
+    #[arg(short = '1', overrides_with_all = ["one", "two", "four", "algorithm"])]
+    one: bool,
+
+    /// use SHA256 for the DS hash
+    #[arg(short = '2', overrides_with_all = ["one", "two", "four", "algorithm"])]
+    two: bool,
+
+    /// use SHA384 for the DS hash
```
Maybe change the algorithm names to be the mnemonics, so that the algorithms in the help output and the arguments for `--algorithm` are the same: `SHA-1`, `SHA-256`, and `SHA-384`. I got confused for a moment when trying to test the algorithms ^^

tertsdiepraam replied:
These arguments are only there for compatibility. So we're out of luck here :)

We could (later) refactor this with separate argument parsing (like I did for notify).

mozzieongit replied:
Yes, the arguments need to stay for compatibility. I was only confused by the help output saying the algorithm is spelled `SHA1`, and then tried using it with `--algorithm` while testing. The mnemonics for the hash algorithms for DS digests are with dashes (e.g. SHA-1) (while the DNSKEY algorithms don't use delimiters (e.g. RSASHA1)).
I'm just trying to avoid confusion for people trying to migrate to using `--algorithm` or `-a`.

```suggestion
    /// use SHA-1 for the DS hash
    #[arg(short = '1', overrides_with_all = ["one", "two", "four", "algorithm"])]
    one: bool,

    /// use SHA-256 for the DS hash
    #[arg(short = '2', overrides_with_all = ["one", "two", "four", "algorithm"])]
    two: bool,

    /// use SHA-384 for the DS hash
```

tertsdiepraam replied:
Ah gotcha! Will fix

mozzieongit commented on src/commands/key2ds.rs:
```diff
@@ -0,0 +1,166 @@
+            | SecAlg::ECDSAP256SHA256 => DigestAlg::SHA256,
+            SecAlg::ECDSAP384SHA384 => DigestAlg::SHA384,
+            SecAlg::ECC_GOST => DigestAlg::GOST,
+            _ => DigestAlg::SHA1,
```
Is using SHA-1 as the fallback a good idea? If new algorithms were to get introduced and we forget to update the above list, they would fallback to using SHA-1, right?

tertsdiepraam replied:
It's the current ldns behaviour: https://github.com/NLnetLabs/ldns/blob/b87b8e1d212b43f225891570ebc5045c6331dd8a/examples/ldns-key2ds.c#L48C1-L48C14

If we want to change it, let's take it to a separate PR, where we can discuss this in depth.

mozzieongit replied:
Ok

tertsdiepraam commented on src/commands/mod.rs:
```diff
@@ -11,6 +12,13 @@ pub enum Command {
+    /// Generate a DS RR from the DNSKEYS in keyfile
+    ///
+    /// The following file will be created: `K<name>+<alg>+<id>.ds`
+    /// The base name `(K<name>+<alg>+<id>` will be printed to stdout
```
This was copied from `ldns-key2ds` so I assumed it had some meaning before I implemented it. But yeah, probably should be removed.

mozzieongit commented on src/commands/mod.rs:
```diff
@@ -11,6 +12,14 @@ pub enum Command {
+    /// Generate a DS RR from the DNSKEYS in keyfile
+    ///
+    /// The following file will be created for each key:
+    /// `K<name>+<alg>+<id>.ds`.The base name `K<name>+<alg>+<id>`
```
Got another typo :P
```suggestion
    /// `K<name>+<alg>+<id>.ds`. The base name `K<name>+<alg>+<id>`
```

ximon18 commented on src/commands/key2ds.rs:
```diff
@@ -0,0 +1,166 @@
+        for entry in zonefile {
+            let entry = entry.map_err(|e| {
+                format!(
+                    "Error while reading public key file \"{}\": {e}",
```
This seems a strange error to me. There was a problem iterating over a zone file but the error is about reading a public key file?

mozzieongit replied:
key2ds is usually used on key files containing only a single dnskey (e.g. generated by keygen). These files are in zonefile format, and you could use a single zone file containing all dnskeys you want to get DS records from. So maybe saying "Error while reading public key *from* file" will keep it ambiguous yet clear enough to apply to both reading from an actual zone file (used as a normal zone file), and reading from a keyfile (that is in zone file format)?

ximon18 replied:
Ah!

tertsdiepraam replied:
I like @mozzieongit's suggestion. Will push a fix for that.

ximon18 commented on src/commands/key2ds.rs:
```diff
@@ -0,0 +1,166 @@
+            let owner = record.owner();
+
+            // Of the records that we see, we only care about DNSKEY records
+            let ScannedRecordData::Dnskey(dnskey) = record.data() else {
```
You could also just test against record.rtype() and above the class, ttl, owner block too.

tertsdiepraam replied:
I'm not sure I follow. I need the `dnskey` value later, so I think the current implementation makes sense? What are you suggesting exactly?

ximon18 replied:
Ah ok, ignore me.

bal-e commented on src/commands/key2ds.rs:
```diff
@@ -0,0 +1,169 @@
+                let sec_alg = sec_alg.to_int();
+                let filename =
+                    format!("K{owner}+{sec_alg:03}+{key_tag:05}.ds");
+                let mut out_file = File::create(&filename).map_err(|e| {
```
Would it make sense to use `File::create_new()` here?

tertsdiepraam replied:
That is, if we have consensus that we want to break compat here, I'd be happy to change it.

mozzieongit replied:
Now that we have different argument parsing for ldns and dnst, and we can use different defaults, we can probably also introduce differences here. Maybe a simple `running_as_ldns: bool` (clap-hidden) member is enough or should it be an enum like `MultiCallMode::{LDNS,DNST}`/`CompatMode::{None/DNST,LDNS}` (or similar)?

mozzieongit replied:
> Now that we have different argument parsing for ldns and dnst, and we can use different defaults, we can probably also introduce differences here. Maybe a simple `running_as_ldns: bool` (clap-hidden) member is enough or should it be an enum like `MultiCallMode::{LDNS,DNST}`/`CompatMode::{None/DNST,LDNS}` (or similar)?

Oh you implemented that already with the force_overwrite

tertsdiepraam replied:
Yep!

tertsdiepraam commented:
Little update: I changed the arguments for the dnst version so that we can use `--force/-f` for overwriting files and turn the ignoring sep option into the more descriptive `--ignore-sep`. I also added file opening/creation into `Env`, which can now be set up with a temp dir. Adds a bit of complexity, but not too much since we already have to use `Env` everywhere and the tests are quite nice if I do say so myself.

mozzieongit commented on Cargo.toml:
```diff
@@ -10,9 +10,16 @@ path = "src/bin/ldns.rs"
+    "bytes",
+    "unstable-validate",
+] }
 lexopt = "0.3.0"
 
 # for implementation of nsec3 hash until domain has it stabilized
-octseq = { version = "0.5.1", features = ["std"] }
+octseq = { version = "0.5.2", features = ["std"] }
```
Will this be an issue when merging the deletion of these lines with the nsec3 pr #6?

tertsdiepraam replied:
Possibly 🤷🏻 I suppose we'll figure it out once we get there

mozzieongit commented on src/commands/key2ds.rs:
```diff
@@ -0,0 +1,497 @@
+            ignore_sep,
+            write_to_stdout,
+            algorithm,
+            // Preventing overwritten files is a dnst feature that is not
```
```suggestion
            // Preventing overwriting files is a dnst feature that is not
```

mozzieongit commented on src/env/fake.rs:
```diff
@@ -53,6 +58,43 @@ impl Env for FakeEnv {
+    where
+        P: AsRef<Path>,
+    {
+        File::create_new(dbg!(self.in_cwd(path)))
```
Is the `dbg!` meant to be there?

tertsdiepraam replied:
Definitely not haha

mozzieongit reviewed (commented):
Apart from the minor comments, it looks good to me. Also, very thorough testing.
